cpp/src/parquet/encoding.cc

```diff
@@ -2095,19 +2122,22 @@ class DictByteArrayDecoderImpl : public DictDecoderImpl<ByteArrayType>,
 
   template <typename BuilderType>
   Status DecodeArrowNonNull(int num_values, BuilderType* builder, int* out_num_values) {
-    constexpr int32_t buffer_size = 2048;
-    int32_t indices_buffer[buffer_size];
-    int values_decoded = 0;
+    constexpr int32_t kBufferSize = 2048;
+    std::vector<int32_t> indices(kBufferSize);
```

Comment:
Ok, I really think we shouldn't apply this change. This is introducing an additional memory allocation and cache misses in the critical path. There's no recursion here. The stack-allocated array was IMO the correct solution.

(also, I don't think anybody reported any stack overflow for the several years that this code path existed)

Reply:
Can we stick with std::array then, it's much more complete and type-safe than a plain C array.

Reply:
We're not using any of the additional functionality. And the type-safety only matters if doing tricks that aren't being done here.